packages/data/src/components/with-dispatch/test/index.js

```diff
@@ -10,224 +11,138 @@ import withDispatch from '../';
+				<Component />
+			</RegistryProvider>
+		);
+		await user.click( screen.getByRole( 'button' ) );
+		expect( firstRegistry.select( 'demo' ).getCount() ).toBe( 1 );
+		expect( secondRegistry.select( 'demo' ).getCount() ).toBe( 1 );
 	} );
 
-	it(
-		'always calls select with the latest state in the handler passed to ' +
-			'the component',
-		() => {
-			const store = registry.registerStore( 'counter', {
-				reducer: ( state = 0, action ) => {
-					if ( action.type === 'update' ) {
-						return action.count;
-					}
-					return state;
-				},
-				actions: {
-					update: ( count ) => ( { type: 'update', count } ),
-				},
-				selectors: {
-					getCount: ( state ) => state,
+	it( 'always calls select with the latest state in the handler passed to the component', async () => {
+		const user = userEvent.setup();
+		const registry = createRegistry();
+		registry.registerStore( 'counter', storeOptions );
+
+		const Component = withDispatch( ( dispatch, ownProps, { select } ) => {
+			return {
+				update: () => {
+					const currentCount = select( 'counter' ).getCount();
+					dispatch( 'counter' ).increment( currentCount + 1 );
 				},
-			} );
-
-			const Component = withDispatch(
-				( _dispatch, ownProps, { select: _select } ) => {
-					const outerCount = _select( 'counter' ).getCount();
-					return {
-						update: () => {
-							const innerCount = _select( 'counter' ).getCount();
-							expect( innerCount ).toBe( outerCount );
-							const actionReturnedFromDispatch = Promise.resolve(
-								_dispatch( 'counter' ).update( innerCount + 1 )
-							);
-							return expect(
-								actionReturnedFromDispatch
-							).resolves.toEqual( {
-								type: 'update',
-								count: innerCount + 1,
-							} );
-						},
-					};
-				}
-			)( ( props ) => <button onClick={ props.update } /> );
-
-			let testRenderer;
-			act( () => {
-				testRenderer = TestRenderer.create(
-					<RegistryProvider value={ registry }>
-						<Component />
-					</RegistryProvider>
-				);
-			} );
-
-			const counterUpdateHandler =
-				testRenderer.root.findByType( 'button' ).props.onClick;
-
-			act( () => {
-				counterUpdateHandler();
-			} );
-			expect( store.getState() ).toBe( 1 );
-
-			act( () => {
-				counterUpdateHandler();
-			} );
-			expect( store.getState() ).toBe( 2 );
-
-			act( () => {
-				counterUpdateHandler();
-			} );
-			expect( store.getState() ).toBe( 3 );
-		}
-	);
+			};
+		} )( ( props ) => <button onClick={ props.update } /> );
+
+		render(
+			<RegistryProvider value={ registry }>
+				<Component />
+			</RegistryProvider>
+		);
+
+		await user.click( screen.getByRole( 'button' ) );
+		expect( registry.select( 'counter' ).getCount() ).toBe( 1 );
+
+		await user.click( screen.getByRole( 'button' ) );
+		expect( registry.select( 'counter' ).getCount() ).toBe( 3 );
+
+		await user.click( screen.getByRole( 'button' ) );
+		expect( registry.select( 'counter' ).getCount() ).toBe( 7 );
```

Comment:
I find this odd. How come from `1` after the first click it becomes `7` after the third click? I'd expect this to stay `1`, `2` and `3`, respectively.

Reply:
Sorry for being unclear - I understand how it works, but I think it might be easy to miss why it's working that way. Might be a good idea to elaborate, either in a comment or in the name of the test.

Reply:
I will add the inline comment.